text/0938-strict-es-module-support.md

```diff
@@ -0,0 +1,247 @@
+---
+stage: accepted
+start-date: 2023-07-19T00:00:00.000Z
+release-date:
+release-versions:
+teams: # delete teams that aren't relevant
+  - cli
+  - framework
+  - learning
+prs:
+  accepted: https://github.com/emberjs/rfcs/pull/938
+project-link:
+---
+
+<!---
+Directions for above:
+
+stage: Leave as is
+start-date: Fill in with today's date, 2032-12-01T00:00:00.000Z
+release-date: Leave as is
+release-versions: Leave as is
+teams: Include only the [team(s)](README.md#relevant-teams) for which this RFC applies
+prs:
+  accepted: Fill this in with the URL for the Proposal RFC PR
+project-link: Leave as is
+-->
+
+# Strict ES Module Support
+
+## Summary
+
+Deprecate `require()` and `define()` and enforce the ES module spec.
+
+## Motivation
+
+Ember has been using ES modules as the official authoring format for many years.
+But we don't actually follow the ES modules spec. We diverge from the spec in
+several important ways:
+
+ - our "modules" can be accessed via the synchronous, dynamic `require()`
+   function. They evaluate the first time someone `require`s them. This behavior
+   is impossible for true ES modules, which support synchronous-and-static
+   inclusion via `import` or asynchronous-and-dynamic inclusion via `import()`,
+   but never synchronous-and-dynamic inclusion.
+
+ - in the spec, trying to import a nonexistent name from another module is an
+   early error -- your module **will never even execute** if it does this. In
+   the current implementation, your code happily runs and receives an undefined
+   local binding.
+
+ - in the spec, these two lines mean very different things:
+    ```js
+    import * as TheModule from "the-module";
+    import TheModule from "the-module";
+    ```
+    In our current implementation, you can sometimes use them interchangeably.
+
+ - In the spec, modules are always read-only. You cannot mutate them from the outside. In our implementation, you can.
+
+This is very bad, because
+ - our code breaks when you try to run it in environments that actually follow the spec.  
+ - refactors of Ember and its build pipeline that would otherwise be 100%
```

Comment:
I super appreciate the callout to, "if we follow the spec", upgrades will be easier and more stable for consumers